modules/core/keeper/msg_server.go

```diff
@@ -711,30 +711,41 @@ func (k Keeper) ChannelUpgradeInit(goCtx context.Context, msg *channeltypes.MsgC
 		return nil, errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module)
 	}
 
-	upgradeSequence, previousVersion, err := k.ChannelKeeper.ChanUpgradeInit(ctx, msg.PortId, msg.ChannelId, chanCap, msg.ProposedUpgradeChannel, msg.TimeoutHeight, msg.TimeoutTimestamp)
+	proposedUpgrade, err := k.extractUpgradeFromMessage(ctx, msg)
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "failed to get upgrade")
+	}
+
+	upgradeSequence, previousVersion, err := k.ChannelKeeper.ChanUpgradeInit(ctx, msg.PortId, msg.ChannelId, proposedUpgrade)
```

Comment:
I feel like this logic may be better encapsulated within 04-channel than at the core msg server. 

E.g. pass fields and timeout and construct the upgrade within there.

We could maybe return the `Upgrade` and reassign the proposed/app negotiated version before passing to `WriteUpgradeInit`

```
channelUpgrade, previousVersion, err := k.Channelkeeper.ChanUpgradeInit(ctx, msg.PortId, msg.ChannelId, msg.UpgradeFields, msg.UpgradeTimeout)
```

Reply:
Yeah I think this is probably a little nicer